Sources/Basic/PathShims.swift

```diff
@@ -19,31 +19,75 @@ import POSIX
+            default:       return .unknown
+        }
+    }
+    /// True if the file system entity is a regular file.
+    var isFile: Bool {
+        return kind == .file
+    }
+    /// True if the file system entity is a directory.
+    var isDirectory: Bool {
+        return kind == .directory
+    }
+    /// True if the file system entity is a symbolic link.
+    var isSymlink: Bool {
+        return kind == .symlink
+    }
+}
+
+/// Returns a structure containing information about the file system entity at `path`, or nil if that path doesn't exist in the file system.  Read, write or execute permission of the file system entity at `path` itself is not required, but all ancestor directories must be searchable.  If they are not, or if any other file system error occurs, this function throws a SystemError.  If `followSymlink` is true and the file system entity at `path` is a symbolic link, it is traversed;  otherwise it is not (any symbolic links in path components other than the last one are always traversed).  If symbolic links are followed and the file system entity at `path` is a symbolic link that points to a non-existent path, then this function returns nil.  By default, `followSymlink` is true.
+public func stat(_ path: AbsolutePath, followSymlink: Bool = true) throws -> libc.stat? {
+    var sbuf = libc.stat()
+    let rv = followSymlink ? stat(path.asString, &sbuf) : lstat(path.asString, &sbuf)
+    guard rv == 0 || errno == ENOENT else { throw SystemError.stat(errno, path.asString) }
+    return rv == 0 ? sbuf : nil
+}
 
-public func isDirectory(_ path: AbsolutePath) -> Bool {
-    // Copied from Utilities.Path.
-    var mystat = stat()
-    let rv = stat(path.asString, &mystat)
-    return rv == 0 && (mystat.st_mode & S_IFMT) == S_IFDIR
+/// Returns true if and only if `path` refers to an existent file system entity.  If `followSymlink` is true, and the last path component is a symbolic link, the result pertains to the destination of the symlink; otherwise it pertains to the symlink itself.  If any file system error other than non-existence occurs, this function throws an error.
+public func exists(_ path: AbsolutePath, followSymlink: Bool = true) throws -> Bool {
+    return try stat(path, followSymlink: followSymlink) != nil
 }
 
-public func isFile(_ path: AbsolutePath) -> Bool {
-    // Copied from Utilities.Path.
-    var mystat = stat()
-    let rv = stat(path.asString, &mystat)
-    return rv == 0 && (mystat.st_mode & S_IFMT) == S_IFREG
+/// Returns true if and only if `path` refers to an existent file system entity and that entity is a regular file.  If `followSymlink` is true, and the last path component is a symbolic link, the result pertains to the destination of the symlink; otherwise it pertains to the symlink itself.  If any file system error other than non-existence occurs, this function throws an error.
+public func isFile(_ path: AbsolutePath, followSymlink: Bool = true) throws -> Bool {
```

Comment:
I agree that it's less convenient, but in the case of an permission error or I/O error (for example), do we really want to just have the error message be "I expected a file and the thing at that path isn't a file".  Seems more honest to report the actual error.  And we are of course treating a missing entry as not-a-file instead of throwing `ENOENT`.